### PR TITLE
adding stats test

### PR DESCRIPTION
**Pull request by abhilashshetty04**

### CODE REVIEW

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

File: io-engine/tests/resource_stats.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    let pool_0_stat: &Vec<IoStats> = &pool_0_stat.get_ref().stats;
+    let pool_1_stat: &Vec<IoStats> = &pool_1_stat.get_ref().stats;
+    let n_stat = nexus_stat.get(0).unwrap();
+    let pool_0_stat = pool_0_stat.get(0).unwrap();
+    let pool_1_stat = pool_1_stat.get(0).unwrap();
+    assert_eq!(n_stat.num_read_ops, 0);
+    assert_eq!(pool_0_stat.num_write_ops, 0);
+    assert_eq!(pool_1_stat.num_write_ops, 0);
+    info!("Stat reset done!");
+    // Run FIO with okay data size.
```
Comment (tiagolobocastro) [inline io-engine/tests/resource_stats.rs]: what does "okay" data size mean?

File: io-engine/tests/resource_stats.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                    .build(),
+            )
+            .with_verbose_err(true)
+            .build(),
+    )
+    .await
+    .unwrap();
+    info!("Fio completed!");
+    // Checks stats after Fio run.
+    // TODO: This is just an initial validtion to see if stats change after FIO
```
Comment (tiagolobocastro) [inline io-engine/tests/resource_stats.rs]:
```suggestion
    // TODO: This is just an initial validation to see if stats change after FIO
```

File: utils/dependencies
Comment (tiagolobocastro) [inline utils/dependencies:1]: Why is this modified?

File: io-engine/tests/resource_stats.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        .await
+        .unwrap();
+
+    let nexus_stat: &Vec<IoStats> = &nexus_stat.get_ref().stats;
+    let pool_0_stat: &Vec<IoStats> = &pool_0_stat.get_ref().stats;
+    let pool_1_stat: &Vec<IoStats> = &pool_1_stat.get_ref().stats;
+    let n_stat = nexus_stat.get(0).unwrap();
+    let pool_0_stat = pool_0_stat.get(0).unwrap();
+    let pool_1_stat = pool_1_stat.get(0).unwrap();
+    assert_eq!(n_stat.num_read_ops, 0);
```
Comment (tiagolobocastro) [inline io-engine/tests/resource_stats.rs]: Why is this 0 after fio?

Comment (abhilashshetty04) [reply io-engine/tests/resource_stats.rs]: Its outdatated now. I think this assert was after reset.

File: io-engine/tests/resource_stats.rs (hunk 1 above)
Comment (abhilashshetty04) [reply io-engine/tests/resource_stats.rs]: Not using it now

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

File: utils/dependencies
Comment (abhilashshetty04) [reply utils/dependencies:1]: Ill check

Comment (abhilashshetty04) [reply utils/dependencies:1]: Removed it

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

**Comment by pchandra19** [comment]
bors try

**Comment by pchandra19** [comment]
bors try

**Comment by pchandra19** [comment]
bors cancel

**Comment by abhilashshetty04** [comment]
bors try

File: io-engine/tests/resource_stats.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+            "ms_0",
+            Binary::from_dbg("io-engine").with_args(vec!["-l", "1"]),
+        )
+        .add_container_bin(
+            "ms_1",
+            Binary::from_dbg("io-engine").with_args(vec!["-l", "2"]),
+        )
+        .add_container_bin(
+            "ms_nex",
+            Binary::from_dbg("io-engine").with_args(vec!["-l", "2"]),
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]:
```suggestion
            Binary::from_dbg("io-engine").with_args(vec!["-l", "3"]),
```

File: io-engine/tests/resource_stats.rs [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        .with_thin(false)
+        .with_size_mb(REPL_SIZE);
+
+    pool_0.create().await.unwrap();
+
+    repl_1.create().await.unwrap();
+    repl_1.share().await.unwrap();
+
+    // Nexus node
+
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]:
```suggestion
```

File: io-engine/tests/resource_stats.rs [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+    let ms_1 = conn.grpc_handle_shared("ms_1").await.unwrap();
+    let ms_nex = conn.grpc_handle_shared("ms_nex").await.unwrap();
+
+    // Node #0
+    let mut pool_0 = PoolBuilder::new(ms_0.clone())
+        .with_name("pool0")
+        .with_new_uuid()
+        .with_malloc("mem0", POOL_SIZE);
+
+    let mut repl_1 = ReplicaBuilder::new(ms_0.clone())
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: Please make it more intuitive by assigning replica ids matching with builder names, e.g. `r0` should be `repl_0`, similarly others

File: io-engine/tests/resource_stats.rs [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+                    .build(),
+            )
+            .with_verbose_err(true)
+            .build(),
+    )
+    .await
+    .unwrap();
+
+    // Validate stats after Fio run.
+
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]:
```suggestion
```

File: io-engine/tests/resource_stats.rs [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+    // Validate num_write_ops of pool_0, pool_2 and pool_1.
+    assert_eq!(pool_0_stat.num_write_ops, pool_1_stat.num_write_ops);
+    assert_eq!(pool_1_stat.num_write_ops, pool_2_stat.num_write_ops);
+
+    // Validate write_latency_ticks of replica and pool
+    if repl_1_stat.stats.clone().unwrap().write_latency_ticks
+        <= pool_0_stat.write_latency_ticks
+    {
+        panic!("replica wr latency less then pool wr latency");
+    }
```
Comment (tiagolobocastro) [inline io-engine/tests/resource_stats.rs]: these could be assert!(), ex:
```suggestion
    assert!(repl_1_stat.stats.as_ref().unwrap().write_latency_ticks > pool_0_stat.write_latency_ticks, "replica wr latency less then pool wr latency")
```

Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: Should the assert hit for `<=` or only `<`
```suggestion
assert!(
    repl_1_stat.stats.clone().unwrap().write_latency_ticks > pool_0_stat.write_latency_ticks,
    "replica wr latency less then pool wr latency"
);
```

File: io-engine/tests/resource_stats.rs [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    let repl_0_stat: &Vec<ReplicaIoStats> = &repl_0_stat.get_ref().stats;
+    let nexus_stat = nexus_stat.get(0).unwrap();
+    let pool_0_stat = pool_0_stat.get(0).unwrap();
+    let pool_1_stat = pool_1_stat.get(0).unwrap();
+    let repl_1_stat = repl_1_stat.get(0).unwrap();
+    let repl_2_stat = repl_2_stat.get(0).unwrap();
+    let repl_0_stat = repl_0_stat.get(0).unwrap();
+
+    // Validate num_read/write_ops reset across resource.
+    assert_eq!(repl_1_stat.stats.clone().unwrap().num_read_ops, 0);
```
Comment (tiagolobocastro) [inline io-engine/tests/resource_stats.rs]: I think all these can be as_ref ?
```suggestion
    assert_eq!(repl_1_stat.stats.as_ref().unwrap().num_read_ops, 0);
```

File: io-engine/tests/resource_stats.rs [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+    assert_eq!(pool_0_stat.num_write_ops, pool_1_stat.num_write_ops);
+    assert_eq!(pool_1_stat.num_write_ops, pool_2_stat.num_write_ops);
+
+    // Validate write_latency_ticks of replica and pool
+    if repl_1_stat.stats.clone().unwrap().write_latency_ticks
+        <= pool_0_stat.write_latency_ticks
+    {
+        panic!("replica wr latency less then pool wr latency");
+    }
+    if repl_2_stat.stats.clone().unwrap().write_latency_ticks
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: modify to use assert as suggested above

File: io-engine/tests/resource_stats.rs [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        <= pool_1_stat.write_latency_ticks
+    {
+        panic!("replica wr latency less then pool wr latency");
+    }
+    if repl_0_stat.stats.clone().unwrap().write_latency_ticks
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: modify to use assert as suggested above

File: io-engine/tests/resource_stats.rs [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+        <= pool_2_stat.write_latency_ticks
+    {
+        panic!("replica wr latency less then pool wr latency");
+    }
+
+    // Validate read_latency_ticks of replica and pool
+    if repl_1_stat.stats.clone().unwrap().read_latency_ticks
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: modify to use assert as suggested above

File: io-engine/tests/resource_stats.rs [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+        <= pool_0_stat.read_latency_ticks
+    {
+        panic!("replica read latency less then pool read latency");
+    }
+    if repl_2_stat.stats.clone().unwrap().read_latency_ticks
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: modify to use assert as suggested above

File: io-engine/tests/resource_stats.rs [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+        <= pool_1_stat.read_latency_ticks
+    {
+        panic!("replica read latency less then pool read latency");
+    }
+    if repl_0_stat.stats.clone().unwrap().read_latency_ticks
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: modify to use assert as suggested above

File: io-engine/tests/resource_stats.rs [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+    {
+        panic!("replica read latency less then pool read latency");
+    }
+
+    // Validate read_latency_ticks of nexus and replica
+    let replica_read_latency =
+        repl_1_stat.stats.clone().unwrap().read_latency_ticks
+            + repl_2_stat.stats.clone().unwrap().read_latency_ticks
+            + repl_0_stat.stats.clone().unwrap().read_latency_ticks;
+    if nexus_stat.read_latency_ticks <= replica_read_latency {
```
Comment (dsharma-dc) [inline io-engine/tests/resource_stats.rs]: modify to use assert as suggested above

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/tests/resource_stats.rs (hunk 8 above)
Comment (abhilashshetty04) [reply io-engine/tests/resource_stats.rs]: They cant be equal also. So should fail if replica_latency <= pool_latency

Comment (abhilashshetty04) [reply io-engine/tests/resource_stats.rs]: I want to assert that they are not equal as well. Ill avoid if

**Comment by abhilashshetty04** [comment]
bors try

**Review by dsharma-dc — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors merge
